.adr/0017-missed-chain-events.md

```diff
@@ -0,0 +1,127 @@
+
+The node could trust the `chainservice` to always deliver `Channel` updates in order. However, putting data sanitation logic closer to the data itself is safer and less prone to duplicated code since `chainservice` is an interface that can have multiple implementations.
+
+### 2. Updating `lastBlockNum`: `chainservice` sends write on chan to `engine`
+
+If the `chainservice` sends a new `blockNum` on a chan the `engine` each time a new block is mined, the `engine` can trigger a write to store to update the `lastBlockNum`. However, in environments where blocks are mined fast, writing to the `store` for every block could have a performance impact especially if the `store` updates are part of the `engine`'s main `run` loop.
+
+## Decision
+
+After considering the aforementioned alternatives, the following design decisions were made.
+
+### Avoid init race conditions that would cause missed events
+
+To ensure no events are missed or processed out of order during the `chainservice` initialization, we can modify the `chainservice` init routine to acquire/release the `eventTracker`'s `mutex` as shown below:
+
+```go
+	// Prevent go routines from processing events before checkForMissedEvents completes
+	ecs.eventTracker.mu.Lock()
+	{
+		ecs.wg.Add(3)
+		go ecs.listenForEventLogs(errChan, eventSub, eventChan, eventQuery)
+		go ecs.listenForNewBlocks(errChan, newBlockSub, newBlockChan)
+		go ecs.listenForErrors(errChan)
+
+		// Search for any missed events emitted while this node was offline
+		err = ecs.checkForMissedEvents(startBlock)
+		if err != nil {
+			return nil, err
+		}
+	}
+	ecs.eventTracker.mu.Unlock()
+```
+
+The `listenForEventLogs` go routine also tries to acquire the `mutex` before updating the `eventTracker.events` queue but will be blocked until `checkForMissedEvents` completes. This means that `listenForEventLogs` will still be able to detect new chain events, but will not add them to the queue or trigger event processing until `checkForMissedEvents` finishes queueing old events.
+
+### 1. Put `Channel` data protection logic in `Channel` class
+
+Add logic to the `Channel` class to protect its data from stale updates rather than relying on an external component (`chainservice`) to only feed it sanitized data. This seems less error-prone and it could be useful to add a `Channel.LastUpdated` struct in case a user wants to know when the channel was last changed. The `chainservice` will still be expected to feed chain events in order, but the added `Channel` logic will act as another layer of protection against faulty data updates.
+
+The `Channel` class should keep track of the `TxIndex` associated with each chain event in case there are multiple events related to a single channel within the same block. Also, an `ethereum.FilterQuery` cannot specify a `TxIndex` so it will be more computationally efficient to check the `TxIndex` in the `Channel` class rather than in the `chainservice` logic.
```

Comment:
>Also, an `ethereum.FilterQuery` cannot specify a `TxIndex` so it will be more computationally efficient to check the `TxIndex` in the `Channel` class rather than in the `chainservice` logic.

While the `ethereum.FilterQuery` doesn't support querying by `TxIndex` I'm not sure why this makes the chain service less computationally efficient? It seems unlikely that you'd be querying by a specific `TxIndex`.

Reply:
Would you be in favor of removing the one sentence you quoted above?

Reply:
I'd be happy to remove it or reword it? I just don't quite follow why checking the `TxIndex` in the `chainservice` is less efficient than if we do so in the channel.